---
title: "Modelling of dropout in scRNAseq data"
author: "Mingdong Liu"
date: "1/11/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r }
library(Seurat, quietly = TRUE)
library(tidyverse, quietly = TRUE)
require(DESeq2, quietly = TRUE)
require(GGally)
require(flextable)
require(ggplot2)
require(ggven)
require(VennDiagram)

hk = c("ACTB", "PGK1","GAPDH","B2M","RPS13","HPRT1","EEF1A1")

emarkers = data.frame(fullname = c("Prostate cancer-associated protein 6/ P501S/ prostein","Prostatic Acid Phosphatase","prostate-specific androgen-regulated transcription factor","prostate specific antigen","prostate cancer antigen 3","alpha-methylacyl-CoA racemase"),
                     gene = c("SLC45A3","ACPP","NKX3.1","KLK3","PCA3","AMACR"))

tgenes = c("STEAP1","FOLH1")


########################################
##    data
#######################################
tuong <- readRDS("W:/devtm/cbd/users/bdecato/Prostate Single Cell/PMID_34936871_fig1.rds")

#get the raw counts
dta1 <- as.matrix(tuong@assays$RNA@counts)

```

## The pattern of dropout with housekeeping genes  


```{r echo=FALSE}
cell_id = colnames(dta1)
allgenes = rownames(dta1)
ann = tuong@meta.data %>%
  select(nCount_RNA, nFeature_RNA, Author.s.cell.type, Author.s.sub.cell.type,Sample.ID, Louvain.clustering, Subject.ID) %>%
  rownames_to_column(var = "cell_id")

dta1 = t(dta1)
dta1 = dta1 %>%
  as.data.frame() %>%
  rownames_to_column(var = "cell_id")

egene_expand = t(combn(emarkers$gene,2)) 
hk_expand = t(combn(hk, 2))

sgenes1 = c(sample(allgenes,size = 100), as.character(egene_expand[,1]), as.character(hk_expand[,1]), tgenes[1])
sgenes2 = c(sample(allgenes,size = 100),as.character(egene_expand[,2]), as.character(hk_expand[,2]), tgenes[2])


dropout = data.frame(g1 = character(),
                     g2 = character(),
                     Sample.ID = character(),
                     do2 = numeric(),
                     do1a = numeric(),
                     do1b = numeric(),
                     do0 = numeric())

for (i in 1:length(sgenes1)){
  dta2 = dta1 %>%
    as.data.frame() %>%
    select(cell_id, all_of(sgenes1[i]), all_of(sgenes2[i])) %>%
    left_join(ann, by = "cell_id") %>%
    rename(g1 = !!sgenes1[i],
           g2 = !!sgenes2[i])
  
  tmp = dta2 %>%
    group_by(Sample.ID) %>%
    summarise(do2 = sum(g1 == 0 & g2 == 0, na.rm = TRUE),
              do1a = sum(g1 == 0 & g2 != 0, na.rm = TRUE),
              do1b = sum(g1 != 0 & g2 == 0, na.rm = TRUE),
              do0 = sum(g1 != 0 & g2 != 0)) %>%
    mutate(g1 = sgenes1[i],
           g2 = sgenes2[i]) %>%
    relocate(g1,.before = Sample.ID) %>%
    relocate(g2,.after = g1)
  
  dropout = bind_rows(dropout,tmp)
  
}

dropout = dropout %>%
  mutate(pct_do2 = 100*do2/(do2 +do1a + do1b + do0),
         pct_do1 = 100*(do1a +do1b)/(do2 +do1a + do1b + do0))


tmp0 = dropout %>%
  summarise(gene = g1,
            do1 = do2 + do1a,
            tot = do0+do1a+do1b+do2,
            sbj = Sample.ID) %>%
  bind_rows(dropout %>% 
              summarise(gene = g2,
                        do1 = do2 + do1b,
                        tot = do0+do1a+do1b+do2,
                        sbj = Sample.ID) ) %>%
  distinct() %>%
  mutate(pct_do = do1/tot) %>%
  filter(gene %in% hk) %>%
  select(gene, sbj, pct_do) %>%
  pivot_wider(names_from = "sbj", values_from = "pct_do")


tmp0 = t(tmp0)
colnames(tmp0) = tmp0[1,]
tmp0 = tmp0 %>%
  as.data.frame() %>%
  slice(-1) %>%
  mutate_all(as.numeric)  

lowerfun <- function(data, mapping) {
  ggplot(data = data, mapping = mapping)+ 
    geom_point(alpha = .25) + 
    geom_smooth(method = "lm", formula = y ~ x, 
                fill = "blue", color = "red", size = 0.5)
}

ggpairs(tmp0, lower = list(continuous = wrap(lowerfun))) +
  theme_bw()



tmpa = dropout %>%
  filter(!g1 %in% hk) %>%
  group_by(Sample.ID) %>%
  summarise(mean_pct_do2 = mean(pct_do2),
            sd_pct_do2 = sd(pct_do2),
            mean_pct_do1 = mean(pct_do1),
            sd_pct_do1 = sd(pct_do1)) 
tmpb = dropout %>%
  filter(g1 %in% hk) %>%
  group_by(Sample.ID) %>%
  summarise(mean_pct_do2 = mean(pct_do2),
            sd_pct_do2 = sd(pct_do2),
            mean_pct_do1 = mean(pct_do1),
            sd_pct_do1 = sd(pct_do1)) 

dropout %>%
  filter(g1 %in% sgenes1[1:20], g2 %in% sgenes2[1:20]) %>%
  ggplot(aes(x = pct_do1, y = pct_do2)) +
  geom_point() +
  theme_bw() +
  facet_grid( g1 ~ g2)


tmpa %>%
  ggplot(aes(x = mean_pct_do1, y = mean_pct_do2)) +
  geom_point(color = "red") +
  geom_point(data = tmpb, aes(x = mean_pct_do1, y = mean_pct_do2), color = "green") +
  theme_bw()

plot(tmpb$mean_pct_do1, tmpb$mean_pct_do2)

tmp1 = dropout %>%
  filter(g1 %in% hk) %>%
  ggplot(aes(x = pct_do1, y = pct_do2, color = Sample.ID)) +
  geom_point() +
  theme_bw() 


tmp1 = dropout %>%
  filter(g1 %in% hk[1]) %>%
  filter(g2 == hk[2])%>%
  pivot_longer(cols = starts_with("do"), names_to = "measure", values_to = "count1")

tmp2 = dropout %>%
  filter(g1 %in% hk[3]) %>%
  filter(g2 == hk[4]) %>%
  pivot_longer(cols = starts_with("do"), names_to = "measure", values_to = "count2")




tmp1 %>%
  mutate(count2 = tmp2$count2) %>%
  ggplot(aes(x = count1, y = count2, color = Sample.ID)) +
  geom_point() +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw() +
  stat_smooth(method = "lm", formula = y ~ x, geom = "smooth") +
  facet_grid(~ measure)

```
  
### by cell types  
  
```{r echo = FALSE}

dta1 = dta1 %>%
  left_join(ann, by = "cell_id")

dropout_byLouvain = data.frame(g1 = character(),
                     g2 = character(),
                     Louvain.clustering = character(),
                     do2 = numeric(),
                     do1a = numeric(),
                     do1b = numeric(),
                     do0 = numeric())

for (i in 1:length(sgenes1)){
  dta2 = dta1 %>%
    as.data.frame() %>%
    select(cell_id, all_of(sgenes1[i]), all_of(sgenes2[i])) %>%
    left_join(ann, by = "cell_id") %>%
    rename(g1 = !!sgenes1[i],
           g2 = !!sgenes2[i])
  
  tmp = dta2 %>%
    group_by(Louvain.clustering) %>%
    summarise(do2 = sum(g1 == 0 & g2 == 0, na.rm = TRUE),
              do1a = sum(g1 == 0 & g2 != 0, na.rm = TRUE),
              do1b = sum(g1 != 0 & g2 == 0, na.rm = TRUE),
              do0 = sum(g1 != 0 & g2 != 0)) %>%
    mutate(g1 = sgenes1[i],
           g2 = sgenes2[i]) %>%
    relocate(g1,.before = Louvain.clustering) %>%
    relocate(g2,.after = g1)
  
  dropout_byLouvain = bind_rows(dropout_byLouvain,tmp)
  
}

dropout_byLouvain = dropout_byLouvain %>%
  mutate(pct_do2 = 100*do2/(do2 +do1a + do1b + do0),
         pct_do1 = 100*(do1a +do1b)/(do2 +do1a + do1b + do0))


tmp0 = dropout_byLouvain %>%
  summarise(gene = g1,
            do1 = do2 + do1a,
            tot = do0+do1a+do1b+do2,
            Louvain.clustering = Louvain.clustering) %>%
  bind_rows(dropout_byLouvain %>% 
              summarise(gene = g2,
                        do1 = do2 + do1b,
                        tot = do0+do1a+do1b+do2,
                        Louvain.clustering = Louvain.clustering) ) %>%
  distinct() %>%
  mutate(pct_do = do1/tot) %>%
  filter(gene %in% hk) %>%
  select(gene, Louvain.clustering, pct_do) %>%
  pivot_wider(names_from = "Louvain.clustering", values_from = "pct_do")


tmp0 = t(tmp0)
colnames(tmp0) = tmp0[1,]
tmp0 = tmp0 %>%
  as.data.frame() %>%
  slice(-1) %>%
  mutate_all(as.numeric)  

lowerfun <- function(data, mapping) {
  ggplot(data = data, mapping = mapping)+ 
    geom_point(alpha = .25) + 
    geom_smooth(method = "lm", formula = y ~ x, 
                fill = "blue", color = "red", size = 0.5)
}

ggpairs(tmp0, lower = list(continuous = wrap(lowerfun))) +
  theme_bw()


```
Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
